src/qnngds/devices/htron.py

```diff
@@ -4,3 +4,95 @@
 
 import phidl.geometry as pg
 from typing import Tuple, Optional
+
+# -*- coding: utf-8 -*-
+"""
+Created on Wed Mar 04 14:33:29 2024
+
+@author: reedf
+
+"""
+
+import numpy as np
+from phidl import Device
+import phidl.geometry as pg
+from qnngds.utilities import PadPlacement, QnnDevice
+from qnngds.geometries import angled_taper
+from typing import Tuple, List, Union, Optional
+
+def planar_hTron(wire_width: Union[int, float]= 0.3,
+                 gate_width: Union[int, float] = 0.1,
+                 channel_width: Union[int, float] = 0.2,
+                 gap: Union[int, float] = 0.02,
+                 gate_length : Union[int, float]= 0.01,
+                 channel_length: Union[int, float] = 0.01,
+                 layer: int = 1
+                 ) -> QnnDevice:
+    """Create a planar hTron
+
+    Parameters
+    -----------------
+    wire_width : int or float
+        Width of routing wires in microns
+    gate_width : int or float
+        Width of superconducting gate in microns
+    channel_width : int or float
+        Width of superconducting channel in microns
+    gap : int or float
+        Spacing between gate and channel in microns
+    gate_length : int or float
+        Length of superconducting gate in microns
+    channel_length : int or float
+        Length of superconducting channel in microns
+    layer: int
+        Layer to draw device on
+
+    Returns
+    -------------
+    HTRON : QnnDevice
+        A QnnDevice containing a single hTron
+
+    """
+
+    HTRON = Device('hTron')
+
+    ports = []
+    for direction,width,length in ((1,channel_width, channel_length), (-1,gate_width, gate_length)):
+        W = Device('wire')
+        constr = W << pg.straight(size=(width, np.max(length - 4*width,0)), layer=layer)
+        constr.center = [0,0]
+        constr.move([direction*(gap/2+width/2),0])
+        taper = angled_taper(wire_width, width, 45, layer=layer)
+        if direction < 0:
+            taper.mirror()
+        taper_lower = W << taper
+        taper_lower.connect(taper_lower.ports[1], constr.ports[2])
+        taper_upper = W << taper
+        taper_upper.mirror()
+        taper_upper.connect(taper_upper.ports[1], constr.ports[1])
+        ports.append(taper_lower.ports[2])
+        ports.append(taper_upper.ports[2])
+        HTRON << W
+
+    HTRON = pg.union(HTRON, layer=layer)
+
+    final_HTRON = QnnDevice('hTron')
```

Comment:
I think I don't fully understand the vision/architecture of the QnnDevice class, but why is there pad information being specified at the device level? I get specifying pads at the circuit level, but I would think devices would be completely agnostic since they may not be connected to any pads at all.

Reply:
If we want generic experiment and place_on_sample functions, then for simple (non-circuit) devices it seems the only opportunity to specify pad routing parameters is at the device level. Previously, these parameters were specified manually at the experiment level and that's why there was a different experiment function for each device. And because there are very different constraints for e.g. wirebond pads vs multiprobe tip contact pads, it's not always possible to be agnostic to that choice either -- for example, a very long nanowire can't fit horizontally under the multiprobe contacts and be routed correctly the same way a short one can be.

It's maybe a bit inelegant design-wise, but in terms of the development experience it means that all the information that has to be manually input for a new device to be added on its own to a layout can be specified in one function instead of across two or three.

Reply:
Maybe we could make a wrapper for the device at the circuit-level, that way there's no requirement that pad/routing information in the device.

Reply:
There are a lot of use cases for devices without a circuit-level function though, my latest layout is primarily test structures composed of a single device

Reply:
Gotcha. Why don't we leave as is, but I'll make an issue and see if I can come up with a more modular way of doing this so that devices don't need pad information.